app/Models/User.php

```diff
@@ -53,15 +55,35 @@ class User extends Authenticatable
 
     protected $table = 'appuser';
 
-    public function isVocabularyEditor() {
-        $userId = $this->id;
-        $vocabularyEditorPrivilegeLabel = 'Edit Vocabulary';
-        $roles = DB::table('appuser')->join('appuser_role_link', function($join) use ($userId) {
-            $join->on('appuser.id', '=', 'appuser_role_link.uid')->where('appuser.id', '=', $userId);
-        })->join('privilege_role_link','privilege_role_link.rid', '=', 'appuser_role_link.rid')->
-            join('privilege', function($join) use ($vocabularyEditorPrivilegeLabel) {
-            $join->on('privilege_role_link.pid', '=', 'privilege.id')->where('privilege.label', '=', $vocabularyEditorPrivilegeLabel);
-        })->select('appuser_role_link.rid')->get();
-        return $roles->count() > 0;
+    /**
+     * The roles that belong to the user.
+     */
+    public function roles(): BelongsToMany
+    {
+        return $this->belongsToMany(Role::class, 'appuser_role_link', 'uid', 'rid');
+    }
+
+    /**
+     * The permissions that belong to the user.
+     */
+    public function getPermissions()
+    {
+        return Permission::whereIn('id', function ($query) {
+            $query->select('pid')
+                ->from('privilege_role_link')
+                ->whereIn('rid', function ($query) {
+                    $query->select('rid')
+                        ->from('appuser_role_link')
+                        ->where('uid', $this->id);
+                });
+        })->get();
+    }
```

Comment:
Seems like this sort of query should normally be a `hasManyThrough`, but given the number of joins this is probably the best way to do it. 👍

Reply:
Unfortunately the many to many nature of the current schema prevents `hasManyThrough` from working and I tried using [this](https://github.com/staudenmeir/eloquent-has-many-deep) package to provide the ability  to have that many to many deep relationship as a default eloquent collection. The issue that kept cropping up was that the database tables don't follow Laravel conventions so none of it would play nicely which meant we ended up with a permissions getter method instead.

Reply:
Oh, neat I hadn't seen that package before. Yeah, not being able to rely on standard naming conventions is tough. But I think you've made it as clean as possible given the legacy table constraints.